use AsyncLocalStorage instead of our home-grown solutions

bengl opened the pull request:
### What does this PR do?
Switches from our custom implementation(s) of AsyncLocalStorage to the one provided by Node.js.

### Motivation
There's no need to maintain our own implementation of AsyncLocalStorage when Node.js provides it for us. Also, switching to stock AsyncLocalStorage will enable the usage of TracingChannel in the future.

### Additional Notes
The comment in the file that selected a storage implementation suggested just using AsyncLocalStorage once it supports triggerAsyncResource(). That said, literally zero of our code uses triggerAsyncResource(), so this is assumed to be historical and no longer relevant.

### Security 
Datadog employees:
- [ ] If this PR touches code that signs or publishes builds or packages, or handles credentials of any kind, I've requested a review from `@DataDog/security-design-and-guidance`.
- [x] This PR doesn't touch any of that.

simon-id reviewed (dismissed):
profiler test failing otherwise LGTM

Qard commented:
Ah, right. Profiler is using those channels to build their code hotspots linkage to trace info. We'll need to coordinate with them to switch that to their own separate use of async_hooks for those lifecycle events. Ideally I'd like to have a better way to do that linking. 🤔

szegedi commented:
We'll obviously have to investigate this, but off the top of your head do you have an inkling what behavioral difference in stock AsyncLocalStorage vs. our current implementation causes the failure?

bengl commented:
@Qard @szegedi I added an async hook in the wall profiler so it can handle that on its own now. The only remaining thing was the `enter` events, which I've added via shimming AsyncLocalStorage.

rochdev commented on packages/dd-trace/src/profiling/profilers/wall.js:
```diff
@@ -76,6 +76,43 @@ function getWebTags (startedSpans, i, span) {
+
+  // TODO this will change in the future. It won't be experimental forever.
+  const needsWrappedCb =
+    !process.execArgv.includes('--experimental-async-context-frame')
```
If a version ships with it enabled by default this will break. Any other way to detect that the new code is running?

bengl replied:
Apart from version ranges (which we won't know in advance), the only other thing would be that the legacy implementation has two (undocumented) properties hanging off it. Would you rather I check for one or both of those properties instead?

szegedi reviewed (approved):
LGTM – thanks for accommodating profiling here!
